controller/state/taskscheduler.go

```diff
@@ -0,0 +1,252 @@
+	s.logNextRunTime(taskID, jobID)
+
+	if s.runNotice != nil {
+		select {
+		case s.runNotice <- newTaskID:
+		default:
+		}
+	}
+}
+
+// createRunableTask generates a new runable (not scheduled) task from a scheduled task
+func (s *stateDB) createRunableTask(task tasks.Task, tag string, runCount int) (string, error) {
+	newTaskID := uuid.New().String()
+	runableTask := task.MakeRunable(newTaskID, runCount)
+	err := s.saveTask(context.Background(), runableTask, tag)
+	if err != nil {
+		return "", err
+	}
+	return newTaskID, nil
+}
+
+// unassignScheduledTask clears task ownership, in DB table only
+func (s *stateDB) unassignScheduledTask(taskID string) error {
+	ctx := context.Background()
+	err := s.transact(ctx, func(tx *sql.Tx) error {
+		_, err := tx.ExecContext(ctx, unassignScheduledTaskSQL, taskID)
+		return err
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// recoverScheduledTasks reads all tasks that are assigned to the scheduler and
+// schedules them.  This should only be called during controller startup to
+// recover tasks that were scheduled during a previous run of the controller.
+func (s *stateDB) recoverScheduledTasks(ctx context.Context) error {
+	var tasklist []tasks.Task
+	err := s.transact(ctx, func(tx *sql.Tx) error {
+		rows, err := tx.QueryContext(ctx, getAllTasksForOwnerSQL, schedulerOwner)
+		if err != nil {
+			return err
+		}
+		defer rows.Close()
+
+		tasklist = nil // reset in case transaction retry
+		for rows.Next() {
+			var serialized string
+			if err = rows.Scan(&serialized); err != nil {
+				return err
+			}
+			tp := tasks.Type.Task.NewBuilder()
+			if err = dagjson.Decoder(tp, bytes.NewBufferString(serialized)); err != nil {
+				return err
+			}
+			task := tp.Build().(tasks.Task)
+			tasklist = append(tasklist, task)
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	for i := range tasklist {
+		err = s.scheduleTask(tasklist[i])
+		if err != nil {
+			log.Errorw("cannot reschedule task", "taskID", tasklist[i].UUID.String(), "err", err)
+		}
+	}
+
+	log.Infow("recovered scheduled tasks", "task_count", len(tasklist))
+	return nil
+}
+
+func (s *stateDB) logNextRunTime(taskID string, jobID cron.EntryID) {
+	ent := s.cronSched.Entry(jobID)
+	if ent.Next.IsZero() {
+		log.Errorw("task is no longer scheduled", "taskID", taskID)
+	}
+	log.Infow("scheduled task", "taskID", taskID, "next_run", ent.Next)
+}
+
+func hasSchedule(task tasks.Task) bool {
+	if t := task.RetrievalTask; t.Exists() {
+		if sch := t.Must().Schedule; sch.Exists() {
+			return sch.Must().String() != ""
+		}
+	}
+	if t := task.StorageTask; t.Exists() {
+		if sch := t.Must().Schedule; sch.Exists() {
+			return sch.Must().String() != ""
+		}
+	}
+	return false
+}
+
+func getTaskSchedule(task tasks.Task) (string, time.Duration) {
+	var schedule, limit string
+	var duration time.Duration
+
+	if t := task.RetrievalTask; t.Exists() {
```

Comment:
goodness this is a sad state of affairs or go-ipld-prime usability :P (not your fault)

btw, you could make this easier by righting a TaskSchedule method over in the tasks module on the Task itself, and then you could use the private members to make it a bit simpler.

Reply:
I like that idea better, will do.